aspnet/fundamentals/servers.rst

```diff
@@ -1,8 +1,274 @@
+
+Kestrel currently supports a limited number of feature interfaces, including ``IHttpRequestFeature``, ``IHttpResponseFeature``, and ``IHttpUpgradeFeature``, but additional features will be added in the future. You can see how these interfaces are implemented and supported by Kestrel in its `ServerRequest class <https://github.com/aspnet/KestrelHttpServer/blob/dev/src/Kestrel/ServerRequest.cs>`_, a portion of which is shown below.
+
+.. code-block:: c#
+	:caption: Kestrel ServerRequest.cs class snippet
+
+	using Microsoft.AspNet.FeatureModel;
+	using Microsoft.AspNet.Http.Features;
+	
+	namespace Kestrel
+	{
+		public class ServerRequest : IHttpRequestFeature, 
+						 IHttpResponseFeature, 
+						 IHttpUpgradeFeature
+		{
+			private FeatureCollection _features;
+			
+			private void PopulateFeatures()
+			{
+				_features.Add(typeof(IHttpRequestFeature), this);
+				_features.Add(typeof(IHttpResponseFeature), this);
+				_features.Add(typeof(IHttpUpgradeFeature), this);
+			}
+		}
+	}
+
+The ``IHttpUpgradeFeature`` interface consists of only one property and one method. Kestrel's implementation of this interface is shown here for reference:
+
+.. code-block:: c#
+	:caption: Kestrel ServerRequest.cs IHttpUpgradeFeature implementation
+	
+	bool IHttpUpgradeFeature.IsUpgradableRequest
+	{
+		get
+		{
+			string[] values;
+			if (_frame.RequestHeaders.TryGetValue("Connection", out values))
+			{
+				return values.Any(value => value.IndexOf("upgrade", 
+					StringComparison.OrdinalIgnoreCase) != -1);
+			}
+			return false;
+		}
+	}
+
+	async Task<Stream> IHttpUpgradeFeature.UpgradeAsync()
+	{
+		_frame.StatusCode = 101;
+		_frame.ReasonPhrase = "Switching Protocols";
+		_frame.ResponseHeaders["Connection"] = new string[] { "Upgrade" };
+		if (!_frame.ResponseHeaders.ContainsKey("Upgrade"))
+		{
+			string[] values;
+			if (_frame.RequestHeaders.TryGetValue("Upgrade", out values))
+			{
+				_frame.ResponseHeaders["Upgrade"] = values;
+			}
+		}
+		_frame.ProduceStart();
+		return _frame.DuplexStream;
+	}
+
+Since Kestrel is open source, it makes an excellent starting point if you need to implement your own custom server. In fact, like all of ASP.NET 5, you're welcome to `contribute <https://github.com/aspnet/KestrelHttpServer/blob/dev/CONTRIBUTING.md>`_ any improvements you make back to the project.
+
+Learn more about working with Kestrel to create :doc:`/tutorials/your-first-mac-aspnet`.
+
+Choosing a server
+-----------------
+
+If you intend to deploy your application on a Windows server, you should prefer the Windows optimized servers (IIS and WebListener). IIS remains the best choice for typical web application deployments on Windows servers, and provides the richest set of features. Choose WebListener instead of IIS only if you need to host your application within your own process or service. Use Kestrel if you intend to deploy to non-Windows environments.
```

Comment:
"Use Kestrel if you intend to develop on or deploy to non-Windows environments."